homeassistant/components/control4/__init__.py

```diff
@@ -135,30 +98,161 @@ async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    except client_exceptions.ClientError as exception:
+        _LOGGER.error("Error connecting to Control4 account API: %s", exception)
+        raise ConfigEntryNotReady from exception
+    except BadCredentials as exception:
+        _LOGGER.error(
+            "Error authenticating with Control4 account API, incorrect username or password: %s",
+            exception,
+        )
+        raise ConfigEntryAuthFailed(exception) from exception
+
+    controller_unique_id = config[CONF_CONTROLLER_UNIQUE_ID]
+    director_token_dict = await account.getDirectorBearerToken(controller_unique_id)
+    no_verify_ssl_session = aiohttp_client.async_get_clientsession(
+        hass, verify_ssl=False
+    )
+
+    director = C4Director(
+        config[CONF_HOST], director_token_dict[CONF_TOKEN], no_verify_ssl_session
+    )
+
+    _LOGGER.debug("Saving new account and director tokens in hass data")
+    entry_data = hass.data[DOMAIN][entry.entry_id]
+    entry_data[CONF_ACCOUNT] = account
+    entry_data[CONF_DIRECTOR] = director
+
+    if not (
+        CONF_WEBSOCKET in entry_data
+        and isinstance(entry_data[CONF_WEBSOCKET], C4Websocket)
+    ):
+        _LOGGER.debug("First time setup, creating new C4Websocket object")
+        connection_tracker = C4WebsocketConnectionTracker(hass, entry)
+        websocket = C4Websocket(
+            config[CONF_HOST],
+            no_verify_ssl_session,
+            connection_tracker.connect_callback,
+            connection_tracker.disconnect_callback,
+        )
+        entry_data[CONF_WEBSOCKET] = websocket
+
+        # Silence C4Websocket related loggers, that would otherwise spam INFO logs with debugging messages
+        logging.getLogger("socketio.client").setLevel(logging.WARNING)
+        logging.getLogger("engineio.client").setLevel(logging.WARNING)
+        logging.getLogger("charset_normalizer").setLevel(logging.ERROR)
+
+    _LOGGER.debug("Starting new WebSocket connection")
+    await entry_data[CONF_WEBSOCKET].sio_connect(director.director_bearer_token)
+
+    _LOGGER.debug(
+        "Registering next token refresh in %s seconds",
+        director_token_dict["validSeconds"],
+    )
+    obj = RefreshTokensObject(hass, entry)
+    async_call_later(
+        hass=hass,
+        delay=director_token_dict["validSeconds"],
+        action=obj.refresh_tokens,
+    )
+
+
+class C4WebsocketConnectionTracker:
+    """Object that provides callables to manually refresh entity states if the Control4 Websocket is disconnected/reconnected."""
+
+    def __init__(self, hass: HomeAssistant, entry: ConfigEntry) -> None:
+        """Initialize the state of the connection tracker object."""
+        self.hass = hass
+        self.entry = entry
+
+        self._was_disconnected = False
+
+    async def connect_callback(self) -> None:
+        """Manually refresh entity states when the Websocket is reconnected after a connection drop."""
+        if self._was_disconnected:
```

Comment:
```suggestion
        if not self._was_disconnected:
```

Reverse the condition, return, and outdent below